### PR TITLE
implement `PySetMethods` and `PyFrozenSetMethods`

### PR DESCRIPTION
**Pull request by davidhewitt**
Split out from #3606. These types are very similar so I've paired them together.

### CODE REVIEW

**Comment by davidhewitt** [comment]
Looks like this need #3695 first for the `Py_LIMITED_API` iterator implementations.

File: src/types/set.rs [hunk 1]
```diff
@@ -71,96 +71,248 @@ impl PySet {
...
+        let py = self.py();
+        inner(self, key.to_object(py).into_bound(py))
     }
 
-    /// Removes and returns an arbitrary element from the set.
-    pub fn pop(&self) -> Option<PyObject> {
-        let element =
-            unsafe { PyObject::from_owned_ptr_or_err(self.py(), ffi::PySet_Pop(self.as_ptr())) };
+    fn pop(&self) -> Option<Bound<'py, PyAny>> {
+        // FIXME should this return PyResult?
```
Comment (adamreichold) [inline src/types/set.rs:222]: From the description in https://docs.python.org/3/c-api/set.html#c.PySet_Pop, I'd say that the current type is right. We could check that the error is indeed a `KeyError` and panic for `SystemError`, i.e. when this was somehow not called on a set, but I think we don't panic due to this elsewhere either.

So while technically we could make this return `PyResult<Option<Bound<PyAny>>>`, this would only catch the invalid situation when `PySet` does reference something that is not actually set and I think this is of limited usefulness.

File: src/types/set.rs [hunk 2]
```diff
@@ -71,96 +71,248 @@ impl PySet {
...
+        Bound::borrowed_from_gil_ref(&self).pop().map(|e| e.into())
+    }
+
+    /// Returns an iterator of values in this set.
+    ///
+    /// # Panics
+    ///
+    /// If PyO3 detects that the set is mutated during iteration, it will panic.
+    pub fn iter(&self) -> PySetIterator<'_> {
+        IntoIterator::into_iter(self)
```
Comment (adamreichold) [inline src/types/set.rs:106]: As with frozenset, I would be glad if we could standardize all iterator related methods to be implemented in terms of the `BoundSetIterator::new` constructor.

File: src/types/frozenset.rs [hunk 3]
```diff
@@ -174,7 +264,7 @@ mod impl_ {
                 if ffi::_PySet_NextEntry(self.set.as_ptr(), &mut self.pos, &mut key, &mut hash) != 0
                 {
                     // _PySet_NextEntry returns borrowed object; for safety must make owned (see #890)
-                    Some(self.set.py().from_owned_ptr(ffi::_Py_NewRef(key)))
+                    Some(key.assume_borrowed(self.set.py()).to_owned())
```
[the remark this answers is not in the document]
Comment (davidhewitt) [reply src/types/frozenset.rs:267]: Agreed, I'm pleased with how these types and names are fitting together!

File: src/types/frozenset.rs [hunk 4]
```diff
@@ -97,46 +97,139 @@ impl PyFrozenSet {
...
+        it: Bound<'p, PyIterator>,
+    }
 
-        fn into_iter(self) -> Self::IntoIter {
-            PyFrozenSetIterator {
-                it: PyIterator::from_object(self).unwrap(),
+    impl<'py> BoundFrozenSetIterator<'py> {
+        pub(super) fn new(frozenset: Bound<'py, PyFrozenSet>) -> Self {
+            Self {
+                it: PyIterator::from_object2(&frozenset).unwrap(),
```
[the remark this answers is not in the document]
Comment (davidhewitt) [reply src/types/frozenset.rs:224]: Yes, I'll rebase that PR once this merges 👍

File: src/types/set.rs [hunk 5]
```diff
@@ -71,96 +71,248 @@ impl PySet {
...
+    pub fn add<K>(&self, key: K) -> PyResult<()>
+    where
+        K: ToPyObject,
+    {
+        Bound::borrowed_from_gil_ref(&self).add(key)
+    }
+
+    /// Removes and returns an arbitrary element from the set.
+    pub fn pop(&self) -> Option<PyObject> {
+        Bound::borrowed_from_gil_ref(&self).pop().map(|e| e.into())
```
[the remark this answers is not in the document]
Comment (davidhewitt) [reply src/types/set.rs:97]: I'll use `Py::from` for now, I think later we might want a `.unbind()` method (`.into_unbound()`?) to go from `Bound` -> `Py`. Not sure what the best name for that is.

The return type here should probably have been `Option<&PyAny>` but I see little point in fixing the gil ref API now.

File: src/types/set.rs [hunk 6]
```diff
@@ -71,96 +71,248 @@ impl PySet {
...
+            ffi::PySet_Clear(self.as_ptr());
+        }
+    }
+
+    #[inline]
+    fn len(&self) -> usize {
+        unsafe { ffi::PySet_Size(self.as_ptr()) as usize }
+    }
+
+    fn is_empty(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (davidhewitt) [reply src/types/set.rs:171]: Can you clarify what you mean, I assume you want me to move this method body into the trait definition?

(I'm not sure there's any material value in doing so; I felt that it made sense to keep all the method bodies together.)

File: src/types/set.rs (hunk 1 above)
Comment (davidhewitt) [reply src/types/set.rs:222]: Yes, agreed 👍. Thanks for thinking about this; I'd forgotten I'd left that FIXME in the code!

**Comment by davidhewitt** [comment]
Thanks for the review; I've actioned all feedback save for the default methods where I'm not entirely sure on the ask.

File: src/types/set.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (adamreichold) [reply src/types/set.rs:173]: Yes, I meant moving the body into the trait.

I think this would make sense because of two reasons:
* The implementation will be the same for all implementors (i.e. if we have future variants of `impl PyFooMethods`. I can't find, but I think we already hit a situation where additional impls were discussed.), so it can be shared.
* The implementation is of a "different kind" than the other ones. It does not really call into the FFI but rather it is a convenience method defined purely via other methods in the trait. So I think it would actually make sense to separate them to avoid adding noise to the block containing the "real" implementations.

File: src/types/frozenset.rs [hunk 7]
```diff
@@ -97,46 +97,143 @@ impl PyFrozenSet {
...
+    type Item = Bound<'py, PyAny>;
+    type IntoIter = BoundFrozenSetIterator<'py>;
+
+    /// Returns an iterator of values in this set.
+    fn into_iter(self) -> Self::IntoIter {
+        BoundFrozenSetIterator::new(self)
+    }
+}
+
+impl<'py> IntoIterator for &'_ Bound<'py, PyFrozenSet> {
```
Comment (adamreichold) [inline src/types/frozenset.rs:206]: Didn't we remove this impl elsewhere as well? Since it just clones the `Bound`, i.e. increases the reference count, this should also be something that user code can just write directly, i.e. `val.clone().into_iter()`?

File: src/types/frozenset.rs [hunk 8]
```diff
@@ -271,6 +372,25 @@ mod tests {
...
+    fn test_frozenset_iter_size_hint() {
+        Python::with_gil(|py| {
+            let set = PyFrozenSet::new(py, &[1]).unwrap();
+
+            let mut iter = set.iter();
+
+            if cfg!(Py_LIMITED_API) {
+                assert_eq!(iter.size_hint(), (0, None));
+            } else {
+                assert_eq!(iter.len(), 1);
```
Comment (adamreichold) [inline src/types/frozenset.rs:385]: This fails to compile as both arms of the if clause are still built even if only one is actually reachable, i.e. this needs to use blocks or maybe even just separate test case definitions.

File: src/types/set.rs [hunk 9]
```diff
@@ -71,96 +71,251 @@ impl PySet {
...
+    ///
+    /// # Panics
+    ///
+    /// If PyO3 detects that the set is mutated during iteration, it will panic.
+    fn into_iter(self) -> Self::IntoIter {
+        BoundSetIterator::new(self)
+    }
+}
+
+impl<'py> IntoIterator for &'_ Bound<'py, PySet> {
```
Comment (adamreichold) [inline src/types/set.rs:286]: Same question as for the `&Bound` impl above.

File: src/types/set.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (adamreichold) [reply src/types/set.rs]: > The return type here should probably have been Option<&PyAny> but I see little point in fixing the gil ref API now.

Very much agree.

Yes, for naming this, since our actual favourite and possible future name for `Py` was `PyDetached`, maybe `.detach()` would already be the right direction and does seem reasonably descriptive within the whole "bind to GIL" picture.

But I am would also be happy with `.unbind()` and I think both would be nicer than `Py::from` which is already nicer than `e.into()`.

File: src/types/frozenset.rs (hunk 7 above)
Comment (davidhewitt) [reply src/types/frozenset.rs]: I feel like it was useful at some point, but I cannot reproduce that now. I've pushed a commit to remove this implementation for now (for all `Bound` types).

File: src/types/set.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (davidhewitt) [reply src/types/set.rs]: Done 👍

File: src/types/set.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (davidhewitt) [reply src/types/set.rs]: Pushed a commit which adds `Bound::unbind`, and maps over that here.

**Review by adamreichold — APPROVED** [review]
